File: cpp/eddsa.cpp
/*
 * Copyright (c) 2012-2020 MIRACL UK Ltd.
 *
 * This file is part of MIRACL Core
 * (see https://github.com/miracl/core).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// EDDSA API functions

// see https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-5.pdf

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include "eddsa_ZZZ.h"

#if CURVETYPE_ZZZ == EDWARDS

using namespace XXX;
using namespace YYY;

// Process a random BIG r by RFC7748 (for Montgomery & Edwards curves only)
static void RFC7748(BIG r)
{
    int c,lg=0;
    BIG t;
    c=ZZZ::CURVE_Cof_I;
    while (c!=1)
    {
        lg++;
        c/=2;
    }
    int n=8*MODBYTES_XXX-lg+1;
    BIG_mod2m(r,n);
    BIG_zero(t); BIG_inc(t,1); BIG_shl(t,n);
    BIG_add(r,r,t);
    c=BIG_lastbits(r,lg);
    BIG_dec(r,c);
//    printf("lg= %d n=%d\n",lg,n);
}

// reverse first n bytes of buff - for little endian
static void reverse(int n,char *buff) {
    for (int i = 0; i < n/2; i++) { 
        char ch = buff[i]; 
        buff[i] = buff[n - i - 1]; 
        buff[n - i - 1] = ch; 
    } 
}

// dom2 - context still needs to be appended
static void dom2(bool ph,int cl,octet *DOM) {
    OCT_jstring(DOM,(char *)"SigZZZ no ZZZ collisions");
    if (ph)
        OCT_jbyte(DOM,1,1);
    else 
        OCT_jbyte(DOM,0,1);
    OCT_jbyte(DOM,cl,1); // context length
}

// dom4 - context still needs to be appended
static void dom4(bool ph,int cl,octet *DOM) {
    OCT_jstring(DOM,(char *)"SigZZZ");
    if (ph)
        OCT_jbyte(DOM,1,1);
    else 
        OCT_jbyte(DOM,0,1);
    OCT_jbyte(DOM,cl,1); // context length
}

static void H(octet *S,char *digest)
{
#if CURVE_SECURITY_ZZZ <= 128  // for ed25519?
    hash512 sh512;
    HASH512_init(&sh512);
    for (int i=0;i<S->len;i++ )
        HASH512_process(&sh512,S->val[i]);
    HASH512_hash(&sh512,digest);
#else                          // for ed448?
    sha3 SHA3;
    SHA3_init(&SHA3,SHAKE256);
    for (int i=0;i<S->len;i++) 
        SHA3_process(&SHA3,S->val[i]);
    SHA3_shake(&SHA3,digest,2*S->len); 
#endif
}

static int H2(bool ph,octet *context,octet *R,octet *Q,octet *M,DBIG dr)
{
    char h[128];
    int b=Q->len;
    char dom[64];
    octet DOM={0,sizeof(dom),dom};
    int cl;
    if (context==NULL) cl=0;
    else cl=context->len&0xff;
#if CURVE_SECURITY_ZZZ <= 128  // for ed25519?
    hash512 sh512;
    HASH512_init(&sh512);
    if (ph || cl>0)
    {                   // if not prehash and no context, omit dom2()
        dom2(ph,cl,&DOM);
        for (int i=0;i<DOM.len;i++)
            HASH512_process(&sh512,DOM.val[i]);
        for (int i=0;i<cl;i++)
            HASH512_process(&sh512,context->val[i]);
    }
    for (int i=0;i<b;i++ )
        HASH512_process(&sh512,R->val[i]);
    for (int i=0;i<b;i++ )
        HASH512_process(&sh512,Q->val[i]);
    for (int i=0;i<M->len;i++ )
        HASH512_process(&sh512,M->val[i]);
    HASH512_hash(&sh512,h);
#else                          // for ed448?
    dom4(ph,cl,&DOM);
    sha3 SHA3;
    SHA3_init(&SHA3,SHAKE256);
    for (int i=0;i<DOM.len;i++)
        SHA3_process(&SHA3,DOM.val[i]);
    for (int i=0;i<cl;i++)
        SHA3_process(&SHA3,context->val[i]);
    for (int i=0;i<b;i++) 
        SHA3_process(&SHA3,R->val[i]);
    for (int i=0;i<b;i++) 
        SHA3_process(&SHA3,Q->val[i]);
    for (int i=0;i<M->len;i++) 
        SHA3_process(&SHA3,M->val[i]);
    SHA3_shake(&SHA3,h,2*b);
#endif
    reverse(2*b,h);
    BIG_dfromBytesLen(dr,h,2*b);
    return 2*b;
}

static void getR(bool ph,int b,char *digest,octet *context,octet *M,DBIG dr) {
    char h[128];
    char dom[64];
    octet DOM={0,sizeof(dom),dom};
    int cl;
    if (context==NULL) cl=0;
    else cl=context->len&0xff;

#if CURVE_SECURITY_ZZZ <= 128  // for ed25519?
    hash512 sh512;
    HASH512_init(&sh512);
    if (ph || cl>0)
    {                   // if not prehash and no context, omit dom2()
        dom2(ph,cl,&DOM);
        for (int i=0;i<DOM.len;i++)
            HASH512_process(&sh512,DOM.val[i]);
        for (int i=0;i<cl;i++)
            HASH512_process(&sh512,context->val[i]);
    }
    for (int i=b;i<2*b;i++ )
        HASH512_process(&sh512,digest[i]);
    for (int i=0;i<M->len;i++)
        HASH512_process(&sh512,M->val[i]);
    HASH512_hash(&sh512,h);

#else                       // for ed448?
    dom4(ph,cl,&DOM);
    sha3 SHA3;
    SHA3_init(&SHA3,SHAKE256);
    for (int i=0;i<DOM.len;i++)
        SHA3_process(&SHA3,DOM.val[i]);
    for (int i=0;i<cl;i++)
        SHA3_process(&SHA3,context->val[i]);
    for (int i=b;i<2*b;i++ )
        SHA3_process(&SHA3,digest[i]);
    for (int i=0;i<M->len;i++) 
        SHA3_process(&SHA3,M->val[i]);
    SHA3_shake(&SHA3,h,2*b); 
#endif

    reverse(2*b,h);
    BIG_dfromBytesLen(dr,h,2*b);
}
// encode integer (little endian)
static int encode_int(XXX::BIG x,char *ei) {
    int b,index=0;

    if (8*MODBYTES_XXX==MBITS_YYY) index=1; // extra byte needed for compression        
    b=MODBYTES_XXX+index;

    ei[0]=0;
    BIG_toBytes(&ei[index],x);
    reverse(b,ei);
    return b;
}

// encode point
static void encode(ZZZ::ECP *P,octet *W) {
    BIG x,y;
    int b,index=0;

    if (8*MODBYTES_XXX==MBITS_YYY) index=1; // extra byte needed for compression        
    b=MODBYTES_XXX+index;

    ECP_get(x, y, P);
    b=encode_int(y,W->val);
    W->val[b-1]|=BIG_parity(x)<<7;
    W->len=b;
}

// decode integer (little endian)
static int decode_int(bool strip_sign,char *ei,XXX::BIG x) {
    int b,index=0;
    int sign=0;
    char r[MODBYTES_XXX+1];

    if (8*MODBYTES_XXX==MBITS_YYY) index=1; // extra byte needed for compression        
    b=MODBYTES_XXX+index;

    for (int i=0;i<b;i++)
        r[i]=ei[i];
  
    reverse(b,r);
    if ((r[0]&0x80)!=0)
        sign=1;
    if (strip_sign)
        r[0]&=0x7f;

    if (r[0]==0) // remove leading zero
        BIG_fromBytesLen(x,&r[1],b-1);
    else
        BIG_fromBytesLen(x,&r[0],b);
    return sign;
}

// decode compressed point
static void decode(octet *W,ZZZ::ECP *P) {
    BIG X,Y;
    FP x,d,t,one,hint;
    int sign=decode_int(true,W->val,Y); // LSB of x
    FP_nres(&x,Y); 
    FP_sqr(&x,&x);
    FP_copy(&d,&x); FP_one(&one);
    FP_sub(&x,&x,&one);
    FP_norm(&x);
    FP_rcopy(&t, ZZZ::CURVE_B);

    FP_mul(&d,&d,&t);
#if CURVE_A_ZZZ == 1
    FP_sub(&d,&d,&one);
#else
    FP_add(&d,&d,&one);
#endif  
    FP_norm(&d);
// inverse square root trick for sqrt(x/d)
    FP_sqr(&t,&x);
    FP_mul(&x,&x,&t);
    FP_mul(&x,&x,&d);
    if (!FP_qr(&x,&hint))
    {
        ECP_inf(P);
        return;
    }
    FP_sqrt(&d,&x,&hint);
    FP_inv(&x,&x,&hint);
    FP_mul(&x,&x,&d);
    FP_mul(&x,&x,&t);

    FP_reduce(&x);
    FP_redc(X,&x);
    if (BIG_parity(X)!=sign)
        FP_neg(&x,&x);
    FP_norm(&x);
    FP_redc(X,&x);
    ECP_set(P,X,Y);
}

/* Calculate a public/private EC GF(p) key pair. Q=D.G mod EC(p),
 * where D is the secret key and Q is the public key
 * and G is fixed generator.
 * RNG is a cryptographically strong RNG 
 * If RNG==NULL, D is generated externally 
 */
int ZZZ::EDDSA_KEY_PAIR_GENERATE(csprng *RNG,octet* D,octet *Q)
{
    BIG r, x, y, s;
    ECP G;
    char digest[128];
    int res = EDDSA_OK;
    int b,index=0;

    if (8*MODBYTES_XXX==MBITS_YYY) index=1; // extra byte needed for compression        
    b=MODBYTES_XXX+index;

    ECP_generator(&G);

    if (RNG != NULL)
        OCT_rand(D, RNG, b); // create random private key

    H(D,digest);

// reverse bytes for little endian
    reverse(b,digest);
    BIG_fromBytes(s,&digest[index]);
    RFC7748(s);
    ECP_mul(&G, s);

    encode(&G,Q);
    return res;
}

// Generate a signature using key pair (D,Q) on message M
// Set ph=true if message has already been pre-hashed
// if ph=false, then context should be NULL for ed25519. However RFC8032 mode ed25519ctx is supported by supplying a non-NULL or non-empty context
int ZZZ::EDDSA_SIGNATURE(bool ph,octet *D, octet *context,octet *M,octet *SIG)
{
    DBIG dr;
    BIG s,sr,sd,q;
    ECP R;
    char digest[128];
    char pq[MODBYTES_XXX+1];
    octet Q={0,sizeof(pq),pq};
    H(D,digest);   // hash of private key
    int res = EDDSA_OK;
    int b,index=0;
    if (8*MODBYTES_XXX==MBITS_YYY) index=1; // extra byte needed for compression        
    b=MODBYTES_XXX+index;

    EDDSA_KEY_PAIR_GENERATE(NULL,D,&Q);

    BIG_rcopy(q, CURVE_Order);
    ECP_generator(&R);

    if (D->len!=Q.len || D->len!=b)
        res=EDDSA_INVALID_PUBLIC_KEY;
    if (res==EDDSA_OK)
    {
        getR(ph,b,digest,context,M,dr);
        BIG_dmod(sr,dr,q);
        ECP_mul(&R,sr);
        encode(&R,SIG);
// reverse bytes for little endian        
        reverse(b,digest);
        BIG_fromBytes(s,&digest[index]);
        RFC7748(s);
        H2(ph,context,SIG,&Q,M,dr);
        BIG_dmod(sd,dr,q);
        BIG_modmul(s,s,sd,q);
        BIG_modadd(s,s,sr,q);
        encode_int(s,&SIG->val[b]);
        SIG->len=2*b;
    }
    return res;
}

// verify a signature using public key Q
// same context (if any) as used for signature
bool ZZZ::EDDSA_VERIFY(bool ph,octet *Q,octet *context,octet *M,octet *SIG) 
{
    DBIG du;
    BIG q,t,su;
    ECP R,QD,G;
    int res = EDDSA_OK;
    bool dr,dq;
    int lg=0;
    int c,b,index=0;
    if (8*MODBYTES_XXX==MBITS_YYY) index=1; // extra byte needed for compression        
    b=MODBYTES_XXX+index;

    c=ZZZ::CURVE_Cof_I;
    while (c!=1)
    {
        lg++;
        c/=2;
    }
    BIG_rcopy(q, CURVE_Order);

    decode(SIG,&R);
    if (ECP_isinf(&R)) return false;
    decode_int(false,&SIG->val[b],t);
    if (BIG_comp(t,q)>=0) return false;
    decode(Q,&QD);
    if (ECP_isinf(&QD)) return false;

    H2(ph,context,SIG,Q,M,du);
    BIG_dmod(su,du,q);
    ECP_generator(&G);
    for (int i=0;i<lg;i++)
    { // use cofactor 2^c
        ECP_dbl(&G); ECP_dbl(&QD); ECP_dbl(&R);
    }

    ECP_neg(&QD);
    ECP_mul2(&G,&QD,t,su);

    if (!ECP_equals(&G,&R)) return false;

    return true;
}

#endif

File: javascript/src/eddsa.js
/*
 * Copyright (c) 2012-2020 MIRACL UK Ltd.
 *
 * This file is part of MIRACL Core
 * (see https://github.com/miracl/core).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* EDDSA API Functions */

var EDDSA = function(ctx) {
    "use strict";

    var EDDSA = {

        INVALID_PUBLIC_KEY: -2,
        ERROR: -3,
        //INVALID: -4,
        EFS: ctx.BIG.MODBYTES,
        EGS: ctx.BIG.MODBYTES,
        SHA512: 64,

        bytestostring: function(b) {
            var s = "",
                len = b.length,
                ch, i;

            for (i = 0; i < len; i++) {
                ch = b[i];
                s += ((ch >>> 4) & 15).toString(16);
                s += (ch & 15).toString(16);
            }

            return s;
        },

        asciitobytes: function(s) {
            var b = [],
                i;

            for (i = 0; i < s.length; i++) {
                b.push(s.charCodeAt(i));
            }

            return b;
        },

        stringtobytes: function(s) {
            var b = [],
                i;

            for (i = 0; i < s.length; i+=2) {
                b.push(parseInt(s.substr(i,2),16));
            }

            return b;
        },

        RFC7748: function(r) {
            var c,lg=0;
            var t = new ctx.BIG(1);
            c=ctx.ROM_CURVE.CURVE_Cof_I;
            while (c!=1)
            {
                lg++;
                c/=2;
            }
            var n=8*this.EGS-lg+1;
            r.mod2m(n);
            t.shl(n);
            r.add(t);
            c=r.lastbits(lg);
            r.dec(c);
        },

    // reverse first n bytes of buff - for little endian
        reverse: function(n,buff) {
            for (var i=0;i<n/2;i++) { 
                var ch = buff[i]; 
                buff[i] = buff[n - i - 1]; 
                buff[n - i - 1] = ch; 
            }
        },
            
    // dom - domain function
        dom: function(pp,ph,cl) {
            var PP=this.asciitobytes(pp);
            var len=PP.length;
            var dom = [];
            for (var i=0;i<len;i++ )
                dom[i]=PP[i];
            if (ph) dom[len]=1;
            else dom[len]=0;
            dom[len+1]=cl;
            return dom;
        },

        H: function(S) {
            if (ctx.ECP.AESKEY<=16) { // for ed25519?
                var sh = new ctx.HASH512();
                for (var i=0;i<S.length;i++)
                    sh.process(S[i]);
                return sh.hash();
            } else {                       // for ed448?
                var digest = [];
                var sh=new ctx.SHA3(ctx.SHA3.SHAKE256);
                for (var i=0;i<S.length;i++)
                    sh.process(S[i]);
                sh.shake(digest,2*S.length);
                return digest;
            }
        },

        H2: function(ph,context,R,Q,M)  {
            var b=Q.length;
            var cl;
            if (context==null) cl=0;
            else cl=context.length; 
            if (ctx.ECP.AESKEY<=16) { // Ed25519??
                var sh = new ctx.HASH512();
                if (ph || cl>0)
                {                   // if not prehash and no context, omit dom2()
                    domain=this.dom("Sig"+ctx.config["CURVE"]+" no "+ctx.config["CURVE"]+" collisions",ph,cl);
                    for (var i=0;i<domain.length;i++)
                        sh.process(domain[i]);
                    for (var i=0;i<cl;i++)
                        sh.process(context[i]);
                }
                for (var i=0;i<b;i++ )
                    sh.process(R[i]);
                for (var i=0;i<b;i++ )
                    sh.process(Q[i]);
                for (var i=0;i<M.length;i++ )
                    sh.process(M[i]);
                var h=sh.hash();
                this.reverse(64,h);
                return ctx.DBIG.fromBytes(h);
            } else {                       // for ed448?
                var domain=this.dom("Sig"+ctx.config["CURVE"],ph,cl);
                var h=[];
                var sh=new ctx.SHA3(ctx.SHA3.SHAKE256);
                for (var i=0;i<domain.length;i++)
                    sh.process(domain[i]);
                for (var i=0;i<cl;i++)
                    sh.process(context[i]);
                for (var i=0;i<b;i++) 
                    sh.process(R[i]);
                for (var i=0;i<b;i++) 
                    sh.process(Q[i]);
                for (var i=0;i<M.length;i++) 
                    sh.process(M[i]);
                sh.shake(h,2*b);
                this.reverse(2*b,h);
                return ctx.DBIG.fromBytes(h);
            }
        },

        getR: function(ph,b,digest,context,M) {
            var domain;
            var cl;
            if (context==null) cl=0;
            else cl=context.length; 

            if (ctx.ECP.AESKEY<=16) { // Ed25519??
                var sh = new ctx.HASH512();
                if (ph || cl>0)
                {                   // if not prehash and no context, omit dom2()
                    domain=this.dom("Sig"+ctx.config["CURVE"]+" no "+ctx.config["CURVE"]+" collisions",ph,cl);
                    for (var i=0;i<domain.length;i++)
                        sh.process(domain[i]);
                    for (var i=0;i<cl;i++)
                        sh.process(context[i]);
                }
                for (var i=b;i<2*b;i++ )
                    sh.process(digest[i]);
                for (var i=0;i<M.length;i++)
                    sh.process(M[i]);
                var h=sh.hash();
                this.reverse(64,h);
                return ctx.DBIG.fromBytes(h);
            } else {                       // for ed448?
                domain=this.dom("Sig"+ctx.config["CURVE"],ph,cl);
                var sh=new ctx.SHA3(ctx.SHA3.SHAKE256);
                var h=[];
                for (var i=0;i<domain.length;i++)
                    sh.process(domain[i]);
                for (var i=0;i<cl;i++)
                    sh.process(context[i]);
                for (var i=b;i<2*b;i++ )
                    sh.process(digest[i]);
                for (var i=0;i<M.length;i++) 
                    sh.process(M[i]);
                sh.shake(h,2*b); 
                this.reverse(2*b,h);
                return ctx.DBIG.fromBytes(h);
            }
        },

// encode integer (little endian)
        encode_int: function(x,w) {
            var b,index=0;
            if (8*ctx.BIG.MODBYTES==ctx.FP.MODBITS) index=1; // extra byte needed for compression        
            b=ctx.BIG.MODBYTES+index;

            w[0]=0;
            x.tobytearray(w,index);
            this.reverse(b,w);
            return b;
        },

// encode point
        encode: function(P,w) {
            var b,index=0;
            if (8*ctx.BIG.MODBYTES==ctx.FP.MODBITS) index=1; // extra byte needed for compression        
            b=ctx.BIG.MODBYTES+index;

            var x=P.getX();
            var y=P.getY();
            this.encode_int(y,w);
            w[b-1]|=x.parity()<<7;
        },

// get sign
        getsign: function(x) {
            var b,index=0;
            if (8*ctx.BIG.MODBYTES==ctx.FP.MODBITS) index=1; // extra byte needed for compression        
            b=ctx.BIG.MODBYTES+index;
            if ((x[b-1]&0x80)!=0)
                return 1;
            else 
                return 0;
        },

// decode integer (little endian)
        decode_int: function(strip_sign,ei) {
            var sign=0;
            var b,index=0;
            if (8*ctx.BIG.MODBYTES==ctx.FP.MODBITS) index=1; // extra byte needed for compression        
            b=ctx.BIG.MODBYTES+index;

            var r=[];

            for (var i=0;i<b;i++)
                r[i]=ei[i];
            this.reverse(b,r);

            if (strip_sign)
                r[0]&=0x7f;

            return ctx.BIG.frombytearray(r,index);
        },

// decode compressed point
        decode: function(W) {
            var P=new ctx.ECP();
            var sign=this.getsign(W); // lsb of x
            var y=this.decode_int(true,W);
            var one=new ctx.FP(1);
            var hint=new ctx.FP();
            var x=new ctx.FP(y); 
            x.sqr(); 
            var d=new ctx.FP(x); 
            x.sub(one);
            x.norm();
            var t=new ctx.FP(0);
            t.rcopy(ctx.ROM_CURVE.CURVE_B);
            d.mul(t);
            if (ctx.ECP.CURVE_A==1) d.sub(one);
            if (ctx.ECP.CURVE_A==-1) d.add(one);
            d.norm();
// inverse square root trick for sqrt(x/d)
            t.copy(x);
            t.sqr();
            x.mul(t);
            x.mul(d);
            if (x.qr(hint)!=1)
            {
                P.inf();
                return P;
            }
            d.copy(x.sqrt(hint));
            x.inverse(hint);
            x.mul(d);
            x.mul(t);
            x.reduce();
            if (x.redc().parity()!=sign)
                x.neg();
            x.norm();
            P.setxy(x.redc(),y);
            return P;
        },

/* Calculate a public/private EC GF(p) key pair. Q=D.G mod EC(p),
 * where D is the secret key and Q is the public key
 * and G is fixed generator.
 * RNG is a cryptographically strong RNG 
 * If RNG==NULL, D is provided externally 
 */
        KEY_PAIR_GENERATE: function(RNG, D, Q) {
            var res = 0;
            var b,index=0;
            if (8*ctx.BIG.MODBYTES==ctx.FP.MODBITS) index=1; // extra byte needed for compression        
            b=ctx.BIG.MODBYTES+index;

            var G=ctx.ECP.generator();

            if (RNG != null) {
                for (var i=0;i<b;i++)
                    D[i]=RNG.getByte();
            }

            var digest=this.H(D);

// reverse bytes for little endian
            this.reverse(b,digest);
            var s=ctx.BIG.frombytearray(digest,index);
            this.RFC7748(s);
            G.copy(G.mul(s));
            this.encode(G,Q);
            return res;
        },

// Generate a signature using key pair (D,Q) on message M
// Set ph=true if message has already been pre-hashed
// if ph=false, then context should be NULL for ed25519. However RFC8032 mode ed25519ctx is supported by supplying a non-NULL or non-empty context
        SIGNATURE: function(ph,D,context,M,SIG) {
            var digest=this.H(D);   // hash of private key
            var res = 0;
            var b,index=0;
            if (8*ctx.BIG.MODBYTES==ctx.FP.MODBITS) index=1; // extra byte needed for compression        
            b=ctx.BIG.MODBYTES+index;

            var S=[];
            var Q=[];

            this.KEY_PAIR_GENERATE(null,D,Q);

            var q = new ctx.BIG();
            q.rcopy(ctx.ROM_CURVE.CURVE_Order);
       
            if (D.length!=Q.length || D.length!=b)
                res=INVALID_PUBLIC_KEY;
            if (res==0)
            {
                var dr=this.getR(ph,b,digest,context,M);
                var sr=dr.mod(q);
                var R=ctx.ECP.generator().mul(sr);
                this.encode(R,S);
                for (var i=0;i<b;i++)
                    SIG[i]=S[i];
// reverse bytes for little endian        
                this.reverse(b,digest);
                var s=ctx.BIG.frombytearray(digest,index);
                this.RFC7748(s);
                dr=this.H2(ph,context,SIG,Q,M);
                var sd=dr.mod(q);
                this.encode_int(ctx.BIG.modadd(sr,ctx.BIG.modmul(s,sd,q),q),S);
                for (var i=0;i<b;i++)
                    SIG[b+i]=S[i];
            }
            return res;
        },

// verify a signature using public key Q
// same context (if any) as used for signature
        VERIFY: function(ph,Q,context,M,SIG) {
            var lg=0;
            var b,index=0;
            if (8*ctx.BIG.MODBYTES==ctx.FP.MODBITS) index=1; // extra byte needed for compression        
            b=ctx.BIG.MODBYTES+index;
            var S=[];

            var c=ctx.ROM_CURVE.CURVE_Cof_I;
            while (c!=1) {
                lg++;
                c/=2;
            }
            var q = new ctx.BIG();
            q.rcopy(ctx.ROM_CURVE.CURVE_Order);
            var R=this.decode(SIG);
            if (R.is_infinity()) return false;
            for (var i=0;i<b;i++)
                S[i]=SIG[b+i];
            var t=this.decode_int(false,S);
            if (ctx.BIG.comp(t,q)>=0) return false;
            var du=this.H2(ph,context,SIG,Q,M);
            var su=du.mod(q);

            var G=ctx.ECP.generator();
            var QD=this.decode(Q); 
            if (QD.is_infinity()) return false;
            QD.neg();
            for (var i=0;i<lg;i++)
            { // use cofactor 2^c
                G.dbl(); QD.dbl(); R.dbl();
            }
        
            if (!G.mul2(t,QD,su).equals(R)) return false;
            return true;
        }

    };

    return EDDSA;
};

if (typeof module !== "undefined" && typeof module.exports !== "undefined") {
    module.exports = {
        EDDSA: EDDSA
    };
}

